lib/js/vdom.js

```diff
@@ -0,0 +1,584 @@
+/*
+ * Virtual DOM integration.
+ */
+// Assumes virtual-dom.js has been loaded.
+
+// Our current ****virtual-dom style**** representation of the VDom.
+
+/* Global variables. */
+
+// loaded: This module should only be loaded once per page; it is an error
+// to load the page more than once. "loaded" is set to true when "runDom" is
+// called for the first time.
+let loaded = false;
+
+// Current VDOM representation of the page. Initially undefined.
+let currentVDom = undefined;
+
+// Event handler AP, for event dispatch.
+var evtHandlerAP = undefined;
+// Running subscription ID.
+var currentSubID = 0;
+
+// Root VDom node.
+var rootNode = undefined;
+
+// Quick-access functions from the virtualDom JS module.
+const h = virtualDom.h;
+const diff = virtualDom.diff;
+const patch = virtualDom.patch;
+const createElement = virtualDom.create;
+
+// Globals for subscriptions
+let animFrameKey = null; // Key for current animation frame handler
+let currentSubscriptions =
+  {windowEvents: {}, intervalEvents: {}, animationFrameHandler: null};
+const windowEvents = {}; // Event name |-> { key: Key, handler: function }
+const intervalEvents = {}; // Key |-> Interval reference
+
+
+// Running element ID.
+let elementID = 0;
+
+
+function genElementID() {
+  const ret = elementID;
+  elementID++;
+  return "elem_" + ret.toString();
+}
+
+/* genEventHandler: Takes a target, a Links variant type of `EventHandler`,
+ * and returns a dictionary of the form { evtName |-> [fn] }, where `fn` is a JS
+ * callback which dispatches a message to the event handler process.
+ *
+ * Inputs: target: { type: "window" } | { type: "element", elementID: id }
+ *         handler: Links EventHandler
+ * Output: { evtName |-> [fn] }
+ */
+function genEventHandler(target, evtHandler) {
+  const evtTag = evtHandler["_label"];
+  const handler = evtHandler["_value"];
+
+  /* Some event handlers unconditionally dispatch the message.
+   * Others may choose not to dispatch a message upon invocation.
+   * These two helpers encode these two modes of dispatch. */
+  const unconditionalDispatch = _makeCont(dispatchMessage);
+  const conditionalDispatch =
+    _makeCont(function(fnRes) {
```

Comment:
One day, we should design a proper runtime API for use in foreign code. Currently the API is ad-hoc and brittle, but it will do for now.